### PR TITLE
Recipe system, does not handle item tags

### PR DESCRIPTION
**Pull request by PierrickMarguin**

### CODE REVIEW

**Review by huntears — CHANGES_REQUESTED** [review]
Please document your code.

File: cubic-server/recipes/CraftingShaped.cpp [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+            for (const auto &item : *items) {
+                if (first_item)
+                    first_item = false;
+                else
+                    std::cout << ", ";
+                std::cout << '\"' << Server::getInstance()->getItemConverter().fromProtocolIdToItem(item) << '\"';
+            }
+            std::cout << ']' << std::endl;
+        }
+        std::cout << " -> " << Server::getInstance()->getItemConverter().fromProtocolIdToItem(this->_result) << " (x" << this->_count << ')' << std::endl;
```
Comment (Trompettesib) [inline cubic-server/recipes/CraftingShaped.cpp:98]: please use a string or a stringstream for all the std::cout
Same as Blasting.cpp

File: cubic-server/recipes/CraftingShapeless.cpp [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    {
+        bool first = true;
+        for (const auto &item : this->_ingredients) {
+            if (first == true)
+                first = false;
+            else 
+                std::cout << "+ ";
+            std::cout << "\"" << Server::getInstance()->getItemConverter().fromProtocolIdToItem(item) << "\" ";
+        }
+        std::cout << "= \"" << Server::getInstance()->getItemConverter().fromProtocolIdToItem(this->_result) << "\" (x" << this->_count << ")" << std::endl;
```
Comment (Trompettesib) [inline cubic-server/recipes/CraftingShapeless.cpp:41]: Same as CraftingShaped.cpp

File: cubic-server/recipes/Recipes.cpp [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    }
+
+    const std::string &Recipe::getGroup(void) const noexcept
+    {
+        return (this->_group);
+    }
+
+    void Recipe::dump(void) const
+    {
+        std::cout << "No data for this recipe." << std::endl;
```
Comment (Trompettesib) [inline cubic-server/recipes/Recipes.cpp:72]: Same as Blasting.cpp

File: cubic-server/recipes/Smelting.cpp [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        this->_ingredient = Server::getInstance()->getItemConverter().fromItemToProtocolId(recipe["ingredient"]["item"].get<std::string>());
+        this->_result = Server::getInstance()->getItemConverter().fromItemToProtocolId(recipe["result"].get<std::string>());
+        this->_experience = recipe["experience"].get<nlohmann::json::number_float_t>();
+        this->_cookingTime = recipe["cookingtime"].get<nlohmann::json::number_unsigned_t>();
+        this->setValidity(true);
+    }
+
+    void Smelting::dump(void) const
+    {
+        std::cout << "\"" << Server::getInstance()->getItemConverter().fromProtocolIdToItem(this->_ingredient) << "\" -> \"" << Server::getInstance()->getItemConverter().fromProtocolIdToItem(this->_result) << "\" (cooking for " << this->_cookingTime << " ticks and get " << this->_experience << " xp)" << std::endl;
```
Comment (Trompettesib) [inline cubic-server/recipes/Smelting.cpp:31]: Same as Blasting.cpp

File: cubic-server/recipes/Smithing.cpp [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+            return;
+        this->_base = Server::getInstance()->getItemConverter().fromItemToProtocolId(recipe["base"]["item"].get<std::string>());
+        this->_addition = Server::getInstance()->getItemConverter().fromItemToProtocolId(recipe["addition"]["item"].get<std::string>());
+        this->_result = Server::getInstance()->getItemConverter().fromItemToProtocolId(recipe["result"]["item"].get<std::string>());
+        this->setValidity(true);
+    }
+
+    void Smithing::dump(void) const
+    {
+        std::cout << "\"" << Server::getInstance()->getItemConverter().fromProtocolIdToItem(this->_base) << "\" + \"" << Server::getInstance()->getItemConverter().fromProtocolIdToItem(this->_addition) << "\" = \"" << Server::getInstance()->getItemConverter().fromProtocolIdToItem(this->_result) << "\"" << std::endl;
```
Comment (Trompettesib) [inline cubic-server/recipes/Smithing.cpp:32]: Same as Blasting.cpp

File: cubic-server/recipes/Smoking.cpp [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        this->_ingredient = Server::getInstance()->getItemConverter().fromItemToProtocolId(recipe["ingredient"]["item"].get<std::string>());
+        this->_result = Server::getInstance()->getItemConverter().fromItemToProtocolId(recipe["result"].get<std::string>());
+        this->_experience = recipe["experience"].get<nlohmann::json::number_float_t>();
+        this->_cookingTime = recipe["cookingtime"].get<nlohmann::json::number_unsigned_t>();
+        this->setValidity(true);
+    }
+
+    void Smoking::dump(void) const
+    {
+        std::cout << "\"" << Server::getInstance()->getItemConverter().fromProtocolIdToItem(this->_ingredient) << "\" -> \"" << Server::getInstance()->getItemConverter().fromProtocolIdToItem(this->_result) << "\" (cooking for " << this->_cookingTime << " ticks and get " << this->_experience << " xp)" << std::endl;
```
Comment (Trompettesib) [inline cubic-server/recipes/Smoking.cpp:31]: Same as Blasting.cpp

File: cubic-server/recipes/StoneCutting.cpp [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+            return;
+        this->_ingredient = Server::getInstance()->getItemConverter().fromItemToProtocolId(recipe["ingredient"]["item"].get<std::string>());
+        this->_result = Server::getInstance()->getItemConverter().fromItemToProtocolId(recipe["result"].get<std::string>());
+        this->_count = recipe["count"].get<nlohmann::json::number_unsigned_t>();
+        this->setValidity(true);
+    }
+
+    void StoneCutting::dump(void) const
+    {
+        std::cout << "\"" << Server::getInstance()->getItemConverter().fromProtocolIdToItem(this->_ingredient) << "\" -> \"" << Server::getInstance()->getItemConverter().fromProtocolIdToItem(this->_result) << "\" (x" << this->_count << ")" << std::endl;
```
Comment (Trompettesib) [inline cubic-server/recipes/StoneCutting.cpp:28]: Same as Blasting.cpp

**Review by Trompettesib — COMMENTED** [review]
Looks good ! some minor things to change

**Review by Trompettesib — DISMISSED** [review]

**Review by STMiki — CHANGES_REQUESTED** [review]

**Review by Trompettesib — APPROVED** [review]

**Review by STMiki — APPROVED** [review]

**Review by Trompettesib — APPROVED** [review]

**Pull request merged**